js/packages/openinference-instrumentation-bedrock-agent-runtime/src/instrumentation.ts

```
@@ -210,6 +248,171 @@ export class BedrockAgentInstrumentation extends InstrumentationBase<Instrumenta
+    return result
+      .then(
+        (response: bedrockAgentRunTime.RetrieveAndGenerateCommandOutput) => {
+          span.setAttributes(extractBedrockRagResponseAttributes(response));
+          span.setStatus({ code: SpanStatusCode.OK });
+          span.end();
+          return response;
+        },
+      )
+      .catch((err: Error) => {
+        span.recordException(err);
+        span.setStatus({ code: SpanStatusCode.ERROR, message: err.message });
+        span.end();
+        throw err;
+      });
+  }
+
+  /**
+   * Handles instrumentation for Bedrock RetrieveCommand (document retrieval only).
+   *
+   * This method starts a span for the Retrieve operation, applies base attributes,
+   * and sets document-level attributes from the response. Errors are recorded and
+   * the span is ended with error status if any occur.
+   *
+   * @param args Arguments passed to the original send method.
+   * @param command The RetrieveCommand instance.
+   * @param original The original send method.
+   * @param client The BedrockAgentRuntimeClient instance.
+   * @returns A Promise resolving to the command output, with tracing instrumentation applied.
+   */
+  private _handleRetrieveCommand(
+    args: Parameters<
+      typeof bedrockAgentRunTime.BedrockAgentRuntimeClient.prototype.send
+    >,
+    command: RetrieveCommand,
+    original: (
+      ...args: Parameters<
+        typeof bedrockAgentRunTime.BedrockAgentRuntimeClient.prototype.send
+      >
+    ) => Promise<bedrockAgentRunTime.RetrieveCommandOutput>,
+    client: unknown,
+  ) {
+    const span = this.oiTracer.startSpan("bedrock.retrieve", {
+      kind: SpanKind.INTERNAL,
+      attributes: safelyExtractBaseRetrieveAttributes(command) ?? undefined,
+    });
+    const result = original.apply(client, args);
+    return result
+      .then((response: bedrockAgentRunTime.RetrieveCommandOutput) => {
+        span.setAttributes(extractBedrockRetrieveResponseAttributes(response));
```

this needs to be safe too otherwise we might wind up throwing an instrumentation error to client, same goes for everything in this file as well, we should always catch and mayb e warn with diag.warn but ultimately swallow errors from attribute extraction, otherwise we might throw it to the user in their app somethign like

```typescript
try {
      span.setAttributes(extractBedrockRetrieveResponseAttributes(response));
      span.setStatus({ code: SpanStatusCode.OK });
    } catch (attributeError) {
      // Handle any errors in attribute extraction separately
      span.recordException(attributeError);
    } finally {
      span.end();
    }

```